### DISCUSSION TITLE
How to expose Jaeger custom metrics?

### DISCUSSION THREAD

**Discussion by totalys** [jaegertracing/jaeger#3288] [Q&A] [answered]
I saw at https://github.com/jaegertracing/jaeger-client-go/blob/master/metrics.go that there are several ready-to-go good metrics available.
I need to export a new custom metric of type gauge example:
` SpanFinishedWithDuration metrics.Gauge ``metric:"custom_traces" tags:"result=ok" help:"Span's duration"`` `

How is the proper way to do this?

**Comment by yurishkuro** [comment, marked as answer]
The mechanism is not designed to be extended with custom metrics. If you want your own span-based metrics you can implement span observer (see https://github.com/jaegertracing/jaeger-client-go/tree/master/rpcmetrics)

**Reply by totalys** [reply]
Would you have some example? 
It looks too cumbersome to write the factory just as the one at github.com/uber/jaeger-lib/metrics/metricstest

**Reply by yurishkuro** [reply]
You only need a factory if you're implementing a different SDK for metrics. If you use Prometheus, there is already a factory for it.

**Reply by totalys** [reply]
Sorry, I'm really lost in this new world.

I couldn't find a way to add my required metric about span duration
`SpanFinishedWithDuration metrics.Gauge ``metric:"custom_traces" tags:"result=ok" help:"Span's duration"`` `

I've tryied this, so far:

```
    ...
    factory := prometheus.New()
    metricsObserver := rcpmetrics.NewObserver(factory, rpcmetrics.DefaultNormalizer)
    reporter := jaeger.NewRemoteReporter(transport.NewHTTPTransport(address))
    tracer, closer := jaeger.NewTracer(
        serviceName, jaeger.NewConstSampler(true), reporter,
        jaeger.TracerOptions.Observer(metricsObserver),
        ... // others non-related properties
    )
```

I'm not sure if I should create a struct that implement some interface and then pass this struct as an argument somewhere... I'm really lost here...

**Reply by yurishkuro** [reply]
```
metricsObserver := rcpmetrics.NewObserver(factory, rpcmetrics.DefaultNormalizer)
```
this is going to use built-in observer. You need to implement your own observer and emit your own metrics inside it.
